src/workerd/io/trace.h

```diff
@@ -477,4 +287,158 @@ inline kj::String truncateScriptId(kj::StringPtr id) {
   return kj::str(truncatedId);
 }
 
+// =======================================================================================
+// Span tracing
+//
+// TODO(cleanup): As of now, this aspect of tracing is actually not related to the rest of this
+//   file. Most of this file defines the interface to feed Trace Workers. Span tracing, however,
+//   is currently designed to feed tracing of the Workers Runtime itself for the benefit of the
+//   developers of the runtime.
+//
+//   However, we might potentially want to give trace workers some access to span tracing as well.
+//   But, that hasn't been designed yet, and it's not clear if that would be based on the same
+//   concept of spans or completely separate. In the latter case, these classes should probably
+//   move to a diffeent header.
+
+class SpanBuilder;
+class SpanObserver;
+
+class Span: public kj::Refcounted {
+  // Read-only interface to a trace span. `Span` objects are delivered to `SpanObserver`s for
+  // recording. To create a `Span`, use a `SpanBuilder`.
+
+public:
+  using TagValue = kj::OneOf<bool, int64_t, double, kj::String>;
+  // TODO(someday): Support binary bytes, too.
+  using TagMap = kj::HashMap<kj::StringPtr, TagValue>;
+  using Tag = TagMap::Entry;
+
+  struct Log {
+    kj::Date timestamp;
+    Tag tag;
+  };
+
+  kj::Maybe<Span&> getParent();
+
+  kj::Date getStartTime();
+  kj::Date getEndTime();
+  kj::StringPtr getOperationName();
+  kj::Duration getDuration();
+  const TagMap& getTags();
+  kj::ArrayPtr<const Log> getLogs();
+
+private:
+  friend class SpanBulider;
+
+  SpanObserver& observer;
+  kj::Maybe<kj::Own<Span>> parent;
+
+  kj::StringPtr operationName;
+  kj::Date startTime;
+  kj::Date endTime;
+  TagMap tags;
+  kj::Vector<Log> logs;
+
+  static constexpr auto MAX_LOGS = 1023;
+  uint droppedLogs = 0;
+  // We set an arbitrary (-ish) cap on log messages for safety. If we drop logs because of this,
+  // we report how many in a final "dropped_logs" log.
+  //
+  // At the risk of being too clever, I chose a limit that is one below a power of two so that
+  // we'll typically have space for one last element available for the "dropped_logs" log without
+  // needing to grow the vector.
+
+public:
+  Span(kj::Badge<SpanBuilder>, Span& parent);
+  Span(kj::Badge<SpanBuilder>, SpanObserver& observer);
+  // Don't use these directly, construct a SpanBuilder instead.
+};
+
+class SpanParent {
+  // An opaque reference to a span which can be used to create child spans. This is typically
+  // passed down from a caller to a callee when the caller wants to allow the callee to create
+  // spans for itself that show up as children of the caller's span, but the caller does not
+  // want to give the callee any other ability to modify the parent span.
+
+public:
+  SpanParent(SpanBuilder& builder);
+
+  SpanParent addRef();
+
+private:
+  kj::Maybe<kj::Own<Span>> span;
+};
+
+class SpanBuilder {
```

Comment:
This is mostly similar to what I wrote, but it's missing something that's vital to support for Jaeger tracing: We need to be able to get _something_ from a SpanBuilder or SpanParent that identifies the span so downstream consumers can cross-reference a builder or parent that exists (or doesn't) with data that's in the observer. That way, we can link an outgoing HTTP request with our span. With this design, you'd return something like a `kj::Maybe<Span&>`, but it feels really weird. You can access a `Span&` from basically anywhere, but it's only done and really usable in one specific context, that being `SpanObserver::onComplete`.

This is kinda what the `Span::Id` was intended for. It encodes in the type system that if you have a `Span`, you can read it because it's done, otherwise you'd have an `Id`, `SpanParent`, or `SpanBuilder`. Every other way of referencing the Span data doesn't give you read access because you shouldn't really be able to read it yet.

Reply:
OK, yes, to put this another way, there are two places where we have to write code that is specific to a particular tracing system and its ID format (e.g. Jaeger): ingress and egress. When a request comes in, we need to extract the parent span ID from a special header. Outgoing subrequests need to embed their parent span ID in a similar header.

I haven't shown in this proposal how these would work. But, I don't think they actually require explicit support in this header. Instead, the ingress and egress adapters would have a direct reference to the JaegerSpanCollector (or whatever) and would be able to query it in order to extract the Jaeger ID for a particular span on egress, or initialize a new span with an incoming ID on ingress. So how that works ends up not actually needing to be represented in `trace.h`.

I like this approach because I don't think it's easy to abstract the notion of span IDs. Jaeger span IDs are 64-bit integers, which seem easy. But in a capability RPC system, you could conceivably imagine span IDs being capabilities, which cannot actually be encoded at all. (In fact, this is exactly how I'd like them to be represented in our process sandboxing system.)

> Every other way of referencing the Span data doesn't give you read access because you shouldn't really be able to read it yet.

In my design, the only place where you end up being able to read a `Span` that isn't done is when it is passed to `SpanObserver::onCreated()`. If we really want to prevent that we could do so in a couple ways... maybe we could pass a `void*` pointer instead of `Span&`, and pass the same pointer to the other methods as well (but also pass `Span&` to `onComplete()`). Or we could perhaps abuse constness by passing `const Span&` to `onCreated()` -- since none of the methods are marked const, you then wouldn't be able to read it. But this may be a confusing way of using "const".

Reply:
> would be able to query it in order to extract the Jaeger ID for a particular span on egress

Query it with what? `SpanBuilder` and `SpanParent` don't expose the underlying `Span&`. Where does the egress function get the `Span&` key from.

> I like this approach because I don't think it's easy to abstract the notion of span IDs. Jaeger span IDs are 64-bit integers, which seem easy. But in a capability RPC system, you could conceivably imagine span IDs being capabilities, which cannot actually be encoded at all. (In fact, this is exactly how I'd like them to be represented in our process sandboxing system.)

Right, but there's not anything that really says `Span::Id` has to be a Jaeger span ID. We can make it opaque and under the hood just implement comparison operations on the underlying pointer to the data, which is ultimately what it was at the end of the life of my original PR.

Reply:
> Query it with what? SpanBuilder and SpanParent don't expose the underlying Span&. Where does the egress function get the Span& key from.

We can add a method for that easily enough.

> Right, but there's not anything that really says Span::Id has to be a Jaeger span ID.

I think that's confusing then... if the type is `Span::Id`, people will expect it represents the Jaeger trace ID. The nice thing about literally using a pointer is it gives people the right starting mental model for what this thing is and what lifetime and uniqueness guarantees it likely provides. I don't think encapsulating this in a separate class buys us anything, but it does add cognitive overhead for people trying to understand the interface. (Especially as these IDs would only even be relevant to SpanCollector and ingress/egress, not to most code that manipulates spans...)

Reply:
Though maybe there's a different approach that makes sense here:

```
class SpanObserver {
public:
  virtual kj::Own<SpanId> newId(SpanId& parent) = 0;
  // Called when a new child span is first allocated, to assign an ID
  // for it.

  virtual void report(SpanId& id, Span& span) = 0;
  // Called to report a completed span for the given ID.
};
```

In this approach, `SpanId` does actually represent the Jaeger span ID (or some other trace system's ID). If there are multiple observers, we ask each one to create an ID for each span -- so the span has multiple IDs.

This approach does a bit more allocation but avoids the need for `SpanObserver` to maintain a `HashMap` and doesn't use raw pointers as keys anywhere. I think it ends up more intuitive. The extra allocation would only be needed when there is at least one observer, of course -- i.e. when tracing is actually in effect.

Reply:
I think in the case of multiple observers it gets a bit weird. Do we allocate another SpanId to hold all the other observer SpanIds? It also requires that every observer have a system to ID spans. But in the case of an observer that just wants to collect spans, it's a bit wasteful.

> if the type is Span::Id, people will expect it represents the Jaeger trace ID.

Are you speaking from your perspective or do you think people will really automatically assume that it refers to a tracing system's span ID that isn't actually referenced by anything in the rest of the runtime? At that point it feels like a documentation issue. The class summary should say, upfront, that the ID is opaque and can only be used as a key to map to some other system or to get the parent span ID.

This is ultimately what we have to do anyway with the pointer system since every method that accepts or returns a `Span&` actually details what state the data is in when you get it. `onCreated` says the only thing you can really call is `Span::getParent()`, `onComplete` says everything is callable, `onDrop` says it's not usable after the method returns. `SpanBuilder` and `SpanParent` will probably have the same sort of things on their methods, but with another type we wouldn't have to. The type itself would be documentation and would automatically prevent anything that wouldn't be right in that context.

Reply:
> Do we allocate another SpanId to hold all the other observer SpanIds?

I imagine each Span would hold an array of SpanIds.

> It also requires that every observer have a system to ID spans. But in the case of an observer that just wants to collect spans, it's a bit wasteful.

They could return a dummy singleton SpanId, or maybe just null.

> Are you speaking from your perspective or do you think people will really automatically assume that it refers to a tracing system's span ID that isn't actually referenced by anything in the rest of the runtime?

Most people familiar with tracing are aware that spans have IDs and will likely expect that the thing called "Span::Id" represents said ID.

> documentation issue

People don't read docs unless they are trying to answer a question. Unfortunately that means that docs are extremely bad at correcting wrong assumptions -- because the person doesn't know that their assumption needs correcting, so won't read.

Reply:
> Unfortunately that means that docs are extremely bad at correcting wrong assumptions -- because the person doesn't know that their assumption needs correcting, so won't read.

And people won't expect that because they have a `Span&` that the data is effectively readable? I mean, it's not _unreadable_, but it's definitely not in the state you might _expect_.

> People don't read docs unless they are trying to answer a question.

If SpanId never exposes a method to get some sort of underlying value out of it like a real jaeger span ID or string or anything, won't users wonder "how do I turn this ID into something useful", read the class definition and realize, "ah, I can't do anything with it directly, I need to do {thing written in this class comment}".

Reply:
@harrishancock @jclee y'all have any opinions on this?

Reply:
I'm planning to pick this up again next week FWIW.

Reply:
> Though maybe there's a different approach that makes sense here:

I went ahead and tried this and it turned out even better than expected:

* It turns out a `SpanId` type isn't even needed. Instead, the `SpanObserver` itself can be an object allocated on a per-span basis. `observer->newChild()` allocates a new `SpanObserver` to handle a child span. This avoids the need for any downcasting of `SpanId` objects inside the observer.
* Conveniently, `Span` objects no longer need to be heap-allocated or refcounted, because the `Span` is passed to the `SpanObserver` only once, upon completion. So it turns out this design does not actually increase allocations as I originally expected.
* In retrospect this design "feels right" to me. It's simpler than any other design proposed so far and it's not really possible to "use wrong" -- the type system enforces everything, I think.

Reply:
Yeah I like this a lot better too. It makes a lot of sense. Though one thing I'm not immediately seeing is how we pull information from in-progress spans for egress like for assigning Jaeger spans to output HTTP requests. Do we expose a ref to the observer for spans, builders, and parents and downcast at runtime?

Reply:
I think that gets a little easier now as we could just add a `kj::Maybe<SpanObserver&> getObserver()` method on `SpanParent`. Then it's up to the caller to downcast it to whatever they expect to extract the bits they need.

Reply:
> @harrishancock @jclee y'all have any opinions on this?

I'm late to the party -- had my head in other issues lately -- but for what it's worth I do like how this is turning out. :)